micropython: ARM/ia32 compiler compliance

niewim19 opened the pull request:
With these changes it is possible to compile and run MicroPython port on ia32 and ARM architecture:
 - architecture specific compilation supporting ia32 and ARM
 - missing libphenix math.h functions that have support in MicoPython listed and passed to compilation
 - MicroPython binary stripping
 - stack and heap size are modifiable at compile time
 - `clean` option support
 - copying MicroPython binary to _root/bin_ and to _prog.stripped_

Patch files were cleaned from unnecessary patches and changed to accommodate arguments passed during compilation time:  
 - heap and stack size control (os.patch)
 - added compilation of missing math.h functions (os.patch)
 - removed `.c` files includes (lib.patch)
 - defined missing math.h functions (lib.patch)
 - removed unnecessary patches (mostly lib.patch)

Patch files will be reduced together with changes made to libphoenix listed in PD-93. 

JIRA: BES-145

kemonats commented on micropython/build.sh:
```diff
@@ -4,41 +4,72 @@ set -e
 
 UPYTH_VER="1.15"
 UPYTH="micropython-${UPYTH_VER}"
+UPYTH_CC_BASE=$(echo "${CC}" | rev | cut -c4- | rev )
```
Maybe `CROSS` or `HOST` variable.

niewim19 replied:
Changed to `CROSS`. Thank you.

nalajcie commented on micropython/build.sh:
```diff
@@ -4,41 +4,72 @@ set -e
+# Architecture specific flags/values
+#
+export PHOENIX_MATH_ABSENT="expm1 log1p asinh acosh atanh erf tgamma lgamma copysign __sin __cos __tan __signbit"
+if [[ ${UPYTH_CC_BASE} == "arm-phoenix-" ]]
```
as @kemonats said - You should probably use `CROSS` variable

niewim19 replied:
Followed kemonats and used `CROSS` variable

nalajcie commented on micropython/build.sh:
```diff
@@ -4,41 +4,72 @@ set -e
+	UPYTH_STACKSZ="2048"
+	UPYTH_HEAPSZ="16384"
+	export LDFLAGS_EXTRA=" -Wl,-q "
+	export CFLAGS_EXTRA="-DUPYTH_STACKSZ=${UPYTH_STACKSZ} -DUPYTH_HEAPSZ=${UPYTH_HEAPSZ} -nostartfiles -fpic -fpie -msingle-pic-base -mno-pic-data-is-text-relative "
```
- `-nostartfiles` should not be needed
- this is probably invalid for `armv7a7-imx6ull` target. I would try to use provided `CFLAGS` (with some filtering?)

niewim19 replied:
As building parameters changed drastically this part should be revised again

nalajcie commented on micropython/build.sh:
```diff
@@ -4,41 +4,72 @@ set -e
+then
+	UPYTH_STACKSZ="2048"
+	UPYTH_HEAPSZ="16384"
+	export LDFLAGS_EXTRA=" -Wl,-q "
```
same as below, probably invalid for 6ull, also, `max-page-size` should be defined for smaller executable. Maybe try to use provided LDFLAGS with some sed/awk to chage it to GCC format?

niewim19 replied:
As building/linking process and parameters changed drastically this part should be revised again

nalajcie commented on micropython/build.sh:
```diff
@@ -4,41 +4,72 @@ set -e
-# FIXME - remove clean
-(cd "${PREFIX_UPYTH_SRC}/mpy-cross" && make clean && make all BUILD="${PREFIX_UPYTH_BUILD}")
-(cd "${PREFIX_UPYTH_SRC}/ports/unix" && make clean && make all)
+mkdir -p "${PREFIX_FS}/root/bin/" && cp ${UPYTH}/ports/unix/micropython "${PREFIX_FS}/root/bin/"
```
- use `PREFIX_ROOTFS`
- use `cp -a` to preserve file attributes and timestamp

niewim19 replied:
Reverted to previous form (where those changes were in place)

nalajcie commented on micropython/build.sh:
```diff
@@ -4,41 +4,72 @@ set -e
+#
+([ -n "$CLEAN" ] && (cd "${UPYTH}/mpy-cross" && make clean) && (cd "${UPYTH}/ports/unix" && make clean))
+(cd "${UPYTH}/mpy-cross" && make all BUILD="${PREFIX_UPYTH_BUILD}" CROSS_COMPILE="${UPYTH_CC_BASE}")
+(cd "${UPYTH}/ports/unix" && make all CROSS_COMPILE="${UPYTH_CC_BASE}" && "${UPYTH_CC_BASE}"strip ${STRIPFLAGS_EXTRA} micropython)
```
install not stripped binary to `PREFIX_PROG` and then strip it into `PREFIX_PROG_STRIPPED` (unstripped binary is needed for debugging or at least `addr2line` to work correctly)

niewim19 replied:
Reverted to previous form (where those changes were already in place)

nalajcie commented on micropython/micropython-1.15-patchset/mpy-cross.patch:
```diff
@@ -1,6 +1,33 @@
++CWARN = -Wall
+ CWARN += -Wextra -Wno-unused-parameter -Wpointer-arith
+-CFLAGS = $(INC) $(CWARN) -std=gnu99 $(CFLAGS_MOD) $(COPT) $(CFLAGS_EXTRA)
++CFLAGS += $(INC) $(CWARN) -std=gnu99 $(CFLAGS_MOD) $(COPT) $(CFLAGS_EXTRA)
```
why do we need to change it to `+=` ? Seems like You're trying to simultaneously use ports-provided `CFLAGS` and `CFLAGS_EXTRA`

nalajcie commented on micropython/micropython-1.15-patchset/os.patch:
```diff
@@ -1,6 +1,6 @@
++++ micropython-phoenix/ports/unix/alloc.c	2021-05-21 10:42:06.565325678 +0200
 @@ -49,21 +49,21 @@
      struct _mmap_region_t *next;
  } mmap_region_t;
```
commented-out code is still here :P

niewim19 replied:
Patches cleared, comments not present now

nalajcie commented on micropython/micropython-1.15-patchset/os.patch:
```diff
@@ -261,13 +288,24 @@ diff -Naur micropython-1.15/ports/unix/Makefile micropython-phoenix/ports/unix/M
 -# Flags to link with pthread library
 -LIBPTHREAD = -lpthread
--
++LDFLAGS = $(LDFLAGS_MOD) $(LDFLAGS_ARCH) -lm $(LDFLAGS_EXTRA)
```
FAIL: changed `+=` to `=` and removed `--gc-sections` in the process

Also: 
 - don't remove `LIBPTHREAD` but set `MICROPY_PY_THREAD` correctly
 - don't remove sections of the Makefile, just patch the `mpconfigport.mk` as expected or even treat it as a config file (so add our own `mpconfigport.mk` to the repo and copy it over)

Edit:
I see that `mpconfigport.mk` is already changed, so half of the changes from this patch should not be needed. Please re-check if they are needed indeed and remove unnecessary ones.

niewim19 replied:
Followed the _Edit_ part of your review message. Patches were adjusted for configs in `mpconfigport.mk`. Thanks

nalajcie commented on micropython/micropython-1.15-patchset/py.patch:
```diff
@@ -1,50 +1,65 @@
+             ASM_MUL_REG_REG(emit->as, REG_ARG_2, reg_rhs);
+             emit_post_push_reg(emit, vtype_lhs, REG_ARG_2);
+-        } else if (MP_BINARY_OP_LESS <= op && op <= MP_BINARY_OP_NOT_EQUAL) {
++        } else if (MP_BINARY_OP_LESS == op && op <= MP_BINARY_OP_NOT_EQUAL) {
```
would be great if You could add some comment with explanation why this is needed (eg. `// PHOENIX CHANGE: needed because "-3 < 5" fails on our system due to singbit() invalid implementation`)

niewim19 replied:
Comment added where i think it was necessary

nalajcie commented on micropython/micropython-1.15-patchset/py.patch:
```diff
@@ -1,50 +1,65 @@
  THIS_MAKEFILE := $(lastword $(MAKEFILE_LIST))
  TOP := $(patsubst %/py/mkenv.mk,%,$(THIS_MAKEFILE))
-+DEBUG = 1
++DEBUG = 0
```
maybe just remove?

niewim19 replied:
Removed. Thanks

nalajcie commented on micropython/micropython-1.15-patchset/py.patch:
```diff
@@ -1,50 +1,65 @@
  PYTHON = python3
  
-+CROSS_COMPILE = i386-pc-phoenix-
++CROSS_COMPILE ?= i386-pc-phoenix-
```
should not be needed

nalajcie commented on micropython/micropython-1.15-patchset/py.patch:
```diff
@@ -86,8 +92,17 @@ diff -Naur micropython-1.15/py/modmath.c micropython-phoenix/py/modmath.c
+ // Whether to use internally defined *printf() functions (otherwise external ones)
+ #ifndef MICROPY_USE_INTERNAL_PRINTF
+-#define MICROPY_USE_INTERNAL_PRINTF (1)
++#define MICROPY_USE_INTERNAL_PRINTF (0)
```
should be provided by define in `CFLAGS` instead of patching the default value

niewim19 replied:
As it is part of `mpconfigport.mk` that is substituted, not patched, this review is outdated

nalajcie reviewed (changes requested):
change in good direction, the patches are smaller but I feel there are still unneeded changes included - please revise them.

Please also fix issues found in `build.sh`

niewim19 commented on micropython/build.sh:
```diff
@@ -4,41 +4,72 @@ set -e
-[ -f "$PREFIX_UPYTH/${UPYTH}.tar.xz" ] || wget https://github.com/micropython/micropython/releases/download/v${UPYTH_VER}/${UPYTH}.tar.xz -P "$PREFIX_UPYTH"
-[ -d "${PREFIX_UPYTH_SRC}" ] || tar xf "$PREFIX_UPYTH/${UPYTH}.tar.xz" -C "$PREFIX_UPYTH_BUILD"
+[ -f "${UPYTH}.tar.xz" ] || wget https://github.com/micropython/micropython/releases/download/v${UPYTH_VER}/${UPYTH}.tar.xz
+[ -d "${UPYTH}" ] || ( echo "Unpacking ${UPYTH}.tar.xz..." && tar xf "${UPYTH}.tar.xz" )
```
Unpacking was reverted back to _build

niewim19 commented on micropython/build.sh:
```diff
@@ -4,41 +4,72 @@ set -e
-# Apply patches
+# Patch
+#
+if ! patch -R -p0 -s -f --dry-run < "${PREFIX_UPYTH_PATCHSET}"/lib.patch >/dev/null 2>&1
```
- patchfiles were split into smaller ones
 - patchfiles enumerated
 - added possibility to substitute some files if it makes configuration easier

niewim19 commented on micropython/build.sh:
```diff
@@ -4,41 +4,72 @@ set -e
+#
+# Clean and build
+#
+([ -n "$CLEAN" ] && (cd "${UPYTH}/mpy-cross" && make clean) && (cd "${UPYTH}/ports/unix" && make clean))
```
Yes it is unnecessary. Removed completely. Thanks

niewim19 commented:
### Initial commit
The initial PR was done on not rebased repository so a lot of @nalajcie changes to `buld.sh` were left behind. Current commit incorporates those changes and more:

### Summary of current PR state:
**MicroPython patching:**
 - patches cleared out of most of unnecessary parts
 - patchfiles split made more atomic
 - patchfiles enumerated
 - added possibility to substitute a file with a full copy instead of patching for better configuration (currently used only for `mpconfigport.mk`). `destinations.cfg` used for specifying path for copied files

**MicroPython building:**
 - compiler specific build added (together with different arguments for MicroPython regarding differences in final hardware specs (like bigger stack for ia32)
 - removed forced `clean` as MicroPython is unpacked in `_build`

**MicroPython linking**
Default option of linking is the one using `gcc`. However due to current state of phoenix arm toolchain there is a workaround that forces linking with `ld` when compiling for arm architecture.

niewim19 commented:
This PR was tested on old toolchain. With new toolchain changes to patches are nedded:
- `06_os_main.patch`:
  - Remove additional `realpath()` function

niewim19 commented on micropython/build.sh:
```diff
@@ -4,13 +4,16 @@ set -e
-PREFIX_UPYTH_PATCHSET="${PREFIX_UPYTH}/${UPYTH}-patchset"
+PREFIX_UPYTH_CONFIG="${PREFIX_UPYTH}/${UPYTH}-config/"
 PREFIX_UPYTH_MARKERS="$PREFIX_UPYTH_BUILD/markers/"
+COPYPATH=""
```
Unused variable. Will be addressed together with suggestions after code review

nalajcie commented on micropython/micropython-1.15-config/patches/10_py_mkrules.patch:
```diff
@@ -0,0 +1,16 @@
+ # Do not pass COPT here - it's *C* compiler optimizations. For example,
+ # we may want to compile using Thumb, but link with non-Thumb libc.
+-	$(Q)$(CC) -o $@ $^ $(LIB) $(LDFLAGS)
++ifndef PHOENIX_PORT_LDFLAGS
```
should not be needed, I've checked that adding `-nostartfiles` to LDFLAGS solves for now the `R_ARM_ABS32` relocations in `.rel.text` (but that should be added as a temporary stop-gap with FIXME until it would be fixed in toolchain).

Moreover: our make-exported LDFLAGS are in `LD` format, it would be nice to be able to translate it in some generic way (for rt1064 we should get something like `-Wl,-z,max-page-size=0x10 -Wl,-q `)

nalajcie commented on micropython/micropython-1.15-config/patches/09_py_ifdefs.patch:
```diff
@@ -0,0 +1,135 @@
+ 
+ #if MICROPY_PY_BUILTINS_FLOAT
+ 
+-#include <math.h>
```
why removing `math.h`, maybe this is the source of warnings?

nalajcie reviewed (commented):
Thanks, the patches and build process looks much better now (and patches are easier to review and understand). I've put some remarks but we can merge this as-is (as it's much better than the current version) and handle the issues in next PRs (or do all changes here if You would prefer OR some middle ground - fix easy ones now, and the more complex ones as a separate PRs).

My general motivation is to:
 - implement what we easily can/should in `libphoenix` (would benefit other ports / projects)
 - have as small patches as possible (for easier upgrade to next micropython versions)

niewim19 commented on micropython/micropython-1.15-config/patches/10_py_mkrules.patch:
```diff
@@ -0,0 +1,16 @@
+diff -Naur micropython-1.15/py/mkrules.mk micropython-phoenix/py/mkrules.mk
```
In of our implementation this is dead code as this function is used only in `alloc.c` in function `ffi_closure_alloc()` which is not compiled due to macro `MICROPY_FORCE_PLAT_ALLOC_EXEC` set to `0` in `micropython-1.15/ports/unix/mpconfigport.h`

niewim19 commented on micropython/micropython-1.15-config/patches/07_os_Makefile.patch:
```diff
@@ -0,0 +1,44 @@
+ test_full: $(PROG) $(TOP)/tests/run-tests.py
+ 	$(eval DIRNAME=ports/$(notdir $(CURDIR)))
+ 	cd $(TOP)/tests && MICROPY_MICROPYTHON=../$(DIRNAME)/$(PROG) ./run-tests.py
+-	cd $(TOP)/tests && MICROPY_MICROPYTHON=../$(DIRNAME)/$(PROG) ./run-tests.py -d thread
```
Thanks, removed.

niewim19 commented:
### Micropython PR is ready for final review. 

From previous commit the following has changed:
 - mmap issue is not a problem as it is a dead code
 - `threads` patch was removed
 - destonations.cfg was removed and substituted with plain `cp` for only one file (but directory `files` remained)
 - <math.h> in objfloat.c was restored back

All libphoenix patches will be done separately

nalajcie commented on micropython/build.sh:
```diff
@@ -21,24 +24,55 @@ mkdir -p "$PREFIX_UPYTH_BUILD" "$PREFIX_UPYTH_MARKERS"
+then
+	export STRIPFLAGS_EXTRA="--strip-all"
+else
+	echo "Error: there is no micropython port for chosen architecture"
```
I would issue a warning, not a direct build breakage, I've tested the `micropython` to work on `imx6ull` target without many issues (only basic manual tests) - just setting LDFLAGS_EXTRA to CFLAGS + `-Wl,-z,max-page-size=0x1000`

Maybe we can use one of the `b_*` functions to output the log in color?

niewim19 replied:
From what I see we export `b_log`, `b_die` and `b_install`. The warning is hardly visible if `b_log` is used. 

The best option would be to use `b_die` which halts the build and prints message in red.

nalajcie commented on micropython/build.sh:
```diff
@@ -21,24 +24,55 @@ mkdir -p "$PREFIX_UPYTH_BUILD" "$PREFIX_UPYTH_MARKERS"
+(cd "${PREFIX_UPYTH_SRC}/mpy-cross" && make all BUILD="${PREFIX_UPYTH_BUILD}" CROSS_COMPILE="${CROSS}")
+(cd "${PREFIX_UPYTH_SRC}/ports/unix" && make all CROSS_COMPILE="${CROSS}")
 
 cp -a "${PREFIX_UPYTH_SRC}/ports/unix/micropython" "$PREFIX_PROG"
```
nah, no unstripped version available, so don't put anything in `PREFIX_PROG` (unless You can get the unstripped version somehow)

niewim19 replied:
Copying to prog removed

nalajcie commented on micropython/build.sh:
```diff
@@ -21,24 +24,55 @@ mkdir -p "$PREFIX_UPYTH_BUILD" "$PREFIX_UPYTH_MARKERS"
+#
+if [ "${TARGET_FAMILY}" = "armv7m7" ]
+then
+	export LDFLAGS_EXTRA=" -Wl,-q ${CFLAGS}"
```
maybe we should set the variables inside IF and then just export it once - outside of IF

Other stuff:
 - when linking by GCC we should always include `CFLAGS` while linking (so this is not `armv7m7`-specific)
 - I think we're missing the correct `-z max-page-size` settings (this at least prevented `micropython` to run on `imx6ull`
 - it probably would be wise to unexport `LDFAGS` and `CFLAGS` and remove patches changing `LDFAGS += ...` to `LDFLAGS = ...`
 -  I would go for exporting necessary `STRIPFLAGS` from `phoenix-rtos-build` in the near future (the same as `CFLAGS`/`LDFLAGS` currently) to be able to remove this if-else altogether.

I tried to come up with some generic way of changing `ld` params syntax to `gcc` params syntax, came up not with generic one, but I think it fits our requirements (so we could remove most of this if-else):
```bash
LDFLAGS_EXTRA="$(echo "$LDFALGS" | sed "s/-z \([a-z=0-9\-]*\)/ -Wl,z,\1/g; s/-q/-Wl,q/g")"
```

but it probably needs some more testing.

niewim19 replied:
- moved `CFLAGS_EXTRA` export outside of _if-else_
 - for some reason your _sed_ command did not work _plug and play_ and I got lost in trying to understand/fix it, so i went for two sed approach
 - one `STRIPFLAGS` export

@nalajcie 
 - It is good idea to just compose new `CFLAGS` and `LDFLAGS` in `build.sh` of upyth instead of patching source code and then provide additional arguments. Should I temporarily save `CFLAGS` and `LDFLAGS` as stated below:
```
CFLAGS_TMP=${CLAGS}
unset CFLAGS
make (...)
export CFLAGS="${CFLAGS_TMP} 
```
and the same for LDFLAGS?

nalajcie replied:
Just to respond Your questions:
 - yes, we should minimize the patches by setting variables in `build.sh`
 - no need to restore the environment, every `build.sh` is executed as a subprocess of phoenix-rtos-ports root `build.sh` - so it doesn't affect parent process in any way

nalajcie commented on micropython/build.sh:
```diff
@@ -23,22 +26,55 @@ mkdir -p "$PREFIX_UPYTH_BUILD" "$PREFIX_UPYTH_MARKERS"
-(cd "${PREFIX_UPYTH_SRC}/mpy-cross" && make clean && make all BUILD="${PREFIX_UPYTH_BUILD}")
-(cd "${PREFIX_UPYTH_SRC}/ports/unix" && make clean && make all)
+(cd "${PREFIX_UPYTH_SRC}/mpy-cross" && make all BUILD="${PREFIX_UPYTH_BUILD}" CROSS_COMPILE="${CROSS}")
+(cd "${PREFIX_UPYTH_SRC}/ports/unix" && make all CROSS_COMPILE="${CROSS}")
```
ok, this is the reason for **no out-of-tree** building currently :P.

nalajcie reviewed (changes requested):
Ok, I would go for some changes in `build.sh` and merge this soon - other patches to be "cleaned" as a part of improving libphoenix PRs :)

niewim19 commented on micropython/build.sh:
```diff
@@ -23,22 +26,55 @@ mkdir -p "$PREFIX_UPYTH_BUILD" "$PREFIX_UPYTH_MARKERS"
+export PHOENIX_MATH_ABSENT="expm1 log1p asinh acosh atanh erf tgamma lgamma copysign __sin __cos __tan __signbit"
+if [[ ${CROSS} == "arm-phoenix-" ]]
+then
+	UPYTH_STACKSZ="4096"
```
WIll this work with macOS compilation?

nalajcie replied:
yeah, this is native bash syntax already used in our scripts, see:
 - this file - line 20
 - https://github.com/phoenix-rtos/phoenix-rtos-project/blob/f6bda073d78971b9f664021aa5e9064716d03b69/_targets/build.project.ia32-generic#L14

niewim19 replied:
_sneaky set-if-unset_ is now implemented in `build.sh`

niewim19 commented on micropython/micropython-1.15-config/patches/07_os_Makefile.patch:
```diff
@@ -0,0 +1,48 @@
+ # If the build directory is not given, make it reflect the variant name.
+ BUILD ?= build-$(VARIANT)
+ 
++# Phoenix libraries
```
Patch removed.

niewim19 commented on micropython/build.sh:
```diff
@@ -21,24 +24,55 @@ mkdir -p "$PREFIX_UPYTH_BUILD" "$PREFIX_UPYTH_MARKERS"
+#
+# Architecture specific flags/values set
+#
+if [ "${TARGET_FAMILY}" = "armv7m7" ]
```
Corrected, thanks.

niewim19 commented:
- Added `README` briefely describing patches
- `LDFLAGS` and `CFLAGS`
  - remove `LDFLAGS` related patch from `07_os_Makefile.patch`
  - unset the `LDFLAGS` and `CFLAGS` variable so that MicroPython compilation can start with blank environment only populated with variables set by `build.sh`
  - unified `LDFLAGS` and `CFLAGS` composing syntax in `build.sh`
- substituted hardcoded values for `DT_DIR` and `DT_REG` with libphoenix counterparts
-  `build.sh` script cleanup

This was left:
```c
+#ifndef MSG_DONTROUTE
+#define MSG_DONTROUTE   0x04 //PHOENIX CHANGE: missing in phoenix
+#endif
```
as it needs to be implemented in libphoenix first

niewim19 commented:
@nalajcie I am giving you a hint (as we discussed some time ago) that progress was made in `CGLAGS`/`LDFLAGS` usage and patches volume

nalajcie commented on micropython/build.sh:
```diff
@@ -21,24 +22,51 @@ mkdir -p "$PREFIX_UPYTH_BUILD" "$PREFIX_UPYTH_MARKERS"
+elif [ "${TARGET_FAMILY}" = "ia32" ]; then
+	STRIPEXP="--strip-all"
+else
+	b_die "ERROR! - No architecture for micropython available"
```
not sure we want to explicitly forbid compilation on other platforms (I've run micropython successfully on `imx6ull` even if it wasn't tested by You :p).

I think we may add a warning message that it's not officially supported or something like that but allow compilation anyway

niewim19 replied:
Added warning message with 5 second wait to give user a chance to see it if it happens

nalajcie commented on micropython/build.sh:
```diff
@@ -21,24 +22,51 @@ mkdir -p "$PREFIX_UPYTH_BUILD" "$PREFIX_UPYTH_MARKERS"
+else
+	b_die "ERROR! - No architecture for micropython available"
+fi
+LDFLAGS_VAR=`echo " ${LDFLAGS}" | sed "s/\s/,/g" | sed "s/,-/ -Wl,-/g"`
```
please add comment regarding what You're doing here (eg. `change LDFLAGS from "ld" params format to "gcc" params - prefixing with -Wl, and changing spaces to colons`)

niewim19 replied:
Added your comment 1:1, thanks :)

nalajcie commented on micropython/build.sh:
```diff
@@ -21,24 +22,51 @@ mkdir -p "$PREFIX_UPYTH_BUILD" "$PREFIX_UPYTH_MARKERS"
+	b_die "ERROR! - No architecture for micropython available"
+fi
+LDFLAGS_VAR=`echo " ${LDFLAGS}" | sed "s/\s/,/g" | sed "s/,-/ -Wl,-/g"`
+
```
please apply shellcheck remark

niewim19 replied:
thanks for pointing it out. Remark applied

nalajcie commented on micropython/build.sh:
```diff
@@ -21,24 +22,51 @@ mkdir -p "$PREFIX_UPYTH_BUILD" "$PREFIX_UPYTH_MARKERS"
+export PHOENIX_MATH_ABSENT="expm1 log1p asinh acosh atanh erf tgamma lgamma copysign __sin __cos __tan __signbit"
+export LDFLAGS_EXTRA="${CFLAGS} $LDFLAGS_VAR"
+export CFLAGS_EXTRA="${CFLAGS} -DUPYTH_STACKSZ=${UPYTH_STACKSZ} -DUPYTH_HEAPSZ=${UPYTH_HEAPSZ} "
+export LDFLAGS=""
```
add comment - "clear original ld-format cflags/ldflags"

niewim19 replied:
comment regarding cflags/ldflags clearing added

nalajcie reviewed (commented):
I've pointed out some minor issues regarding comments in `build.sh`, other than that I think we should merge it soon and than incrementally add missing parts to `libphoenix` removing them from patches at the same time (the end-goal is to have minimal or no patches at all ;)).
